components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+    ) -> Result<()> {
+        // First, insert rows for all merged descendants.
+        sql_support::each_sized_chunk(
+            &descendants,
+            sql_support::default_max_variable_number() / 4,
+            |chunk, _| -> Result<()> {
+                // We can't avoid allocating here, since we're binding four
+                // parameters per descendant. Rust's `SliceConcatExt::concat`
+                // is semantically equivalent, but requires a second allocation,
+                // which we _can_ avoid.
+                let mut params = Vec::with_capacity(chunk.len() * 4);
+                for d in chunk.iter() {
+                    params.push(
+                        d.merged_node
+                            .merge_state
+                            .local_node()
+                            .map(|node| node.guid.as_str()),
+                    );
+                    params.push(
+                        d.merged_node
+                            .merge_state
+                            .remote_node()
+                            .map(|node| node.guid.as_str()),
+                    );
+                    params.push(Some(d.merged_node.guid.as_str()));
+                    params.push(Some(d.merged_parent_node.guid.as_str()));
+                }
+                self.db.execute(&format!(
+                    "
+                    INSERT INTO mergedTree(localGuid, remoteGuid, mergedGuid, mergedParentGuid, level,
+                                           position, useRemote, shouldUpload)
+                    VALUES {}",
+                    sql_support::repeat_display(chunk.len(), ",", |index, f| {
+                        let d = &chunk[index];
+                        write!(f, "(?, ?, ?, ?, {}, {}, {}, {})",
+                            d.level, d.position, d.merged_node.merge_state.should_apply(),
+                            d.merged_node.merge_state.upload_reason() != UploadReason::None)
+                    })
+                ), &params)?;
+                Ok(())
+            },
+        )?;
+
+        // Next, insert rows for deletions. Unlike Desktop, there's no
+        // `noteItemRemoved` trigger on `itemsToRemove`, since we don't support
+        // observer notifications.
+        sql_support::each_chunk(&deletions, |chunk, _| -> Result<()> {
+            self.db.execute(
+                &format!(
+                    "
+                    INSERT INTO itemsToRemove(guid, localLevel, shouldUploadTombstone)
+                    VALUES {}",
+                    sql_support::repeat_display(chunk.len(), ",", |index, f| {
+                        let d = &chunk[index];
+                        write!(f, "(?, {}, {})", d.local_level, d.should_upload_tombstone)
+                    })
+                ),
+                chunk.iter().map(|d| d.guid.as_str()),
+            )?;
+            Ok(())
+        })?;
+
+        // "Deleting" from `itemsToMerge` fires the `insertNewLocalItems` and
```

Comment:
Explain air quotes (`itemsToMerge` is a view, so we don't really delete from it).